pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,329 @@
+                          Integer64ToString(getpid()),
+                          handler_);
+}
+
+pid_t ControllerManager::ControllerPid() {
+  GoogleString controller_pid_filename(ControllerPidFilename());
+  if (controller_pid_filename.empty()) {
+    return -1;
+  }
+
+  return ReadPidFromFile(controller_pid_filename);
+}
+
+void ControllerManager::KillControllerIfAlreadyRunning() {
+  pid_t controller_pid = ControllerPid();
+  if (controller_pid == -1) {
+    // Lookup failure, already logged.
+    return;
+  }
+
+  handler_->Message(
+      kInfo,
+      "About to send SIGKILL to babysitter process (%d) to tell it to "
+      "exit.  The controller will see it exit and exit as well.",
+      controller_pid);
+  // Babysitter has no cleanup to do, no reason to avoid SIGKILL.
+  int ret = kill(controller_pid, SIGKILL);
+  if (ret == -1) {
+    // Signal failed to send.
+    if (errno == ESRCH) {
+      // This is fine.  This is what usually happens if the controller exits
+      // abruptly and leaves behind a stale pid in its file.
+    } else if (errno == EPERM) {
+      handler_->Message(
+          kWarning,
+          "Permissions error trying to kill controller process at PID %d.",
+          controller_pid);
+    } else {
+      LOG(DFATAL) << "Got unexpected failure " << errno << " from kill.";
+    }
+    return;  // If the signal didn't go out, no point waiting for an effect.
+  }
+
+  for (int i = 0; i < 100; i++) {
+    // Wait for up to a second for it to die.
+    if (!IsProcessRunning(controller_pid)) {
+      return;
+    }
+    timer_->SleepUs(10 * 1000 /* 0.01s */);
+  }
+  handler_->Message(
+      kWarning, "Couldn't get controller process (%d) to exit", controller_pid);
+}
+
+void ControllerManager::RunController() {
+  handler_->Message(kInfo, "Controller running with PID %d", getpid());
+
+  // Would set up gRPC server here and pass control to its event loop.  Instead
+  // just hang out sleeping forever.
+  while (true) {
+    sleep(60);  // 1m
+  }
+}
+
+void ControllerManager::Daemonize() {
+  // Make a new session (process group).
+  if (setsid() < 0) {
+    handler_->Message(kWarning, "Daemonize: Failed to setsid().");
+  }
+
+  // We need to fork again to make sure there is no session group leader.
+  pid_t pid = fork();
+  CHECK(pid != -1) << "Couldn't fork to daemonize.";
+  if (pid != 0) {
+    exit(EXIT_SUCCESS);
+  }
+
+  // If we keep the current directory we might keep them from being able to
+  // unmount their filesystem.
+  if (chdir("/") < 0) {
+    handler_->Message(kWarning, "Daemonize: Failed to chdir(/).");
+  }
+
+  // If we disconnect file descriptors then logging will break, so don't.
+}
+
+void ControllerManager::ForkOffControllerProcess() {
+  handler_->Message(kInfo, "Forking controller process off of %d", getpid());
+
+  pid_t pid = fork();
+  CHECK(pid != -1) << "Couldn't fork a controller babysitter process";
+
+  if (pid != 0) return;  // Parent process.
```

Comment:
Oh?  The style guide is happy with this case: https://google.github.io/styleguide/cppguide.html#Conditionals